marabunta/core.py

```diff
@@ -52,44 +89,33 @@ def migrate(config):
     database = Database(config)
-    with database.connect() as conn:
-        with conn.cursor() as cursor:
-            try:
-                # If the migration is run concurrently (in several
-                # containers, hosts, ...), only 1 is allowed to proceed
-                # with the migration. It will be the first one to win
-                # the advisory lock. The others will be flagged as 'replica'.
-                replica = False
-                while not pg_advisory_lock(cursor, ADVISORY_LOCK_IDENT):
-                    if not replica:  # print only the first time
-                        safe_print('A concurrent process is already '
-                                   'running the migration')
-                    replica = True
-                    time.sleep(0.5)
-                else:
-                    # when a replica could finally acquire a lock, it
-                    # means that the main process has finished the
-                    # migration. In that case, the replica should just
-                    # exit because the migration already took place. We
-                    # wait till then to be sure we won't run Odoo before
-                    # the main process could finish the migration.
-                    if replica:
-                        return
-
-                table = MigrationTable(cursor)
-                runner = Runner(config, migration, cursor, table)
-                runner.perform()
-            except:
-                # We want to commit even if we had an error in a migration
-                # script! Thus we have an entry in the migration table
-                # indicating that a migration has been started but not
-                # finished
-                conn.commit()
-                raise
-            else:
-                conn.commit()
-            finally:
-                conn.close()
+
+    with database.connect() as connection:
+        application_lock = ApplicationLock(connection)
+        application_lock.start()
+
+        while not application_lock.acquired:
+            time.sleep(0.5)
+        else:
+            if application_lock.replica:
+                # when a replica could finally acquire a lock, it
+                # means that the main process has finished the
+                # migration. In that case, the replica should just
+                # exit because the migration already took place. We
+                # wait till then to be sure we won't run Odoo before
+                # the main process could finish the migration.
+                application_lock.stop = True
+                application_lock.join()
+                return
+            # we are not in the replica: go on for the migration
+
+        try:
+            table = MigrationTable(connection)
+            runner = Runner(config, migration, connection, table)
+            runner.perform()
+        finally:
+            application_lock.stop = True
+            application_lock.join()
 
```

Comment:
@cyrilgdn did you observed an unexpected behavior? There is nothing to commit in the `application_lock` thread because it writes nothing, it only acquires a lock which is released at the closing of the cursor. Besides, I'm pretty sure that the context manager `with self.connection.cursor() as cursor:` commits at the closing.